crates/sui-framework/packages/sui-framework/sources/k_of_n_upgrade_policy.move

```diff
@@ -0,0 +1,522 @@
+            id: proposal_uid,
+            cap: cap_id,
+            signer,
+            digest,
+            allowed_voters,
+            current_voters: vec_set::empty(),
+        })
+    }
+
+    /// Vote in favor of an upgrade, aborts if the ballot is not for the proposed
+    /// upgrade or if the upgrade was already performed.
+    public fun vote(
+        proposal: &mut ProposedUpgrade, 
+        ballot: Ballot,
+        ctx: &TxContext,
+    ) {
+        assert!(proposal.signer != @0x0, EAlreadyIssued);
+        let Ballot { 
+            id, 
+            owner: _,
+            transfers_count: _,
+            digest,
+            proposed_upgrade,
+        } = ballot;
+        assert!(digest == proposal.digest, EInvalidBallot);
+        let proposal_id = object::id(proposal);
+        assert!(proposal_id == proposed_upgrade, EInvalidBallot);
+        let ballot_id = object::uid_to_inner(&id);
+        vec_set::insert(&mut proposal.current_voters, ballot_id);
+        object::delete(id);
+
+        event::emit(UpgradeVoted {
+            proposal: object::id(proposal),
+            digest,
+            ballot: ballot_id,
+            voter: tx_context::sender(ctx),
+        });
+    }
+
+    /// Issue an `UpgradeTicket` for the upgrade being voted on.  Aborts if 
+    /// there are not enough votes yet, or if the upgrade was already performed.
+    /// The signer of the transaction must be the same as the one proposing the
+    /// upgrade.
+    public fun authorize_upgrade(
+        cap: &mut KofNUpgradeCap,
+        proposal: &mut ProposedUpgrade, 
+        ctx: &TxContext,
+    ): UpgradeTicket {
+        assert!(proposal.cap == object::id(cap), EInvalidProposalForUpgrade);
+        assert!(
+            vec_set::size(&proposal.current_voters) >= cap.required_votes, 
+            ENotEnoughVotes,
+        );
+
+        // assert the signer is the proposer and the upgrade has not happened yet
+        let signer = tx_context::sender(ctx);
+        assert!(proposal.signer != @0x0, EAlreadyIssued);
+        assert!(proposal.signer == signer, ESignerMismatch);
+        proposal.signer = @0x0;
+
+        event::emit(UpgradePerformed {
+            upgrade_cap: proposal.cap,
+            proposal: object::id(proposal),
+            digest: proposal.digest,
+            signer,
+        });
+
+        let policy = package::upgrade_policy(&cap.cap);
+        package::authorize_upgrade(
+            &mut cap.cap,
+            policy,
+            proposal.digest,
+        )
+    }
+
+    /// Finalize the upgrade to produce the given receipt.
+    public fun commit_upgrade(
+        cap: &mut KofNUpgradeCap, 
+        receipt: UpgradeReceipt,
+    ) {
+        package::commit_upgrade(&mut cap.cap, receipt)
+    }
+
+    /// Transfer a `Ballot` to the given `address`. 
+    /// Record the fact a transfer was performed.
+    public fun transfer(ballot: Ballot, receiver: address, ctx: &TxContext) {
```

Comment:
Not opposed to it, but this pattern stands a little bit apart from others on Sui (making this cap `key` only and with a custom transfer rule that includes tracking).  I like the fact that this thing can't be wrapped -- that's good for discoverability, but not sure about the rest (given that, it should be fairly easy to track what's happening to it without the events and the count tracking).

Reply:
I have to say I am not sure I follow and we'll talk about this (as the rest) offline too.
The general idea behind upgrade policy was to track as much as possible and as redundantly as possible because the operation is rare and controlled by a policy. Adding the `VoterCap`, for instance, I would love to track transfers (so create custom transfers) for that too.
It seems a sound idea to me to build redundancy here.
But I may have not grasped the concern (as I feel I do for other comments)